rmw_dds_common/include/rmw_dds_common/qos.hpp

```diff
@@ -224,6 +226,28 @@ RMW_DDS_COMMON_PUBLIC
 rmw_qos_profile_t
 qos_profile_update_best_available_for_services(const rmw_qos_profile_t & qos_profile);
 
+/// Parse USER_DATA key=value;key=value; encoding, find value of key "typehash"
+/**
+ * \param[in] user_data USER_DATA qos raw bytes
+ * \param[in] user_data_size Length of user_data
+ * \return Parsed type hash
+ * \return zero-initialized value if data unparseable or key not found
+ */
+RMW_DDS_COMMON_PUBLIC
+rosidl_type_hash_t
+parse_type_hash_from_user_data(
+  const uint8_t * user_data,
+  size_t user_data_size);
```

Comment:
Does this method (and the one below) really belong in the `qos.hpp` header and implementation file?  From what I can tell they don't anything to do with QoS, but I could be mistaken.

Reply:
My thinking is that they are solely for encoding/reading the USER_DATA QoS string format. This encoding is specific to QoS profiles, and has no defined use outside of them. However it's not manipulating a QoS Profile object directly, just the contents of one of its policies. The problem is that each of the 3 implementations has a different type for UserData, so we can't pass something like that in. And, USER_DATA is not part of the ROS 2 QoS API, it's used in DDS QoS to distribute these type hashes (and other information, right now just Service metadata).

I personally think it best belongs here, but if you feel strongly that it should go somewhere else then that's fine too. I could even just rewrite it in each RMW implementation, leaving it as a fully implementation-dependent detail. Though I think it'd be better to keep things in "common" and skip that potential interop problems between DDS impls.

Reply:
OK, that's fine with me.  Just double-checking.  No action to take here.